File: 백준/Silver/15663. N과 M （9）/N과 M （9）.py

import sys

input = sys.stdin.readline

N, M = list(map(int, input().split()))

nums = list(map(int, input().split()))
nums.sort()

result = []
memo = {}

def backtrack(idx, curr): # idx: 사용한 인덱스 리스트
    if len(curr) == M:
        if tuple(curr) not in memo:
            result.append(curr[:])
            memo[tuple(curr)] = True
        return
    
    for i in range(len(nums)):
        if i not in idx:
            curr.append(nums[i])
            idx.append(i)
            backtrack(idx, curr)
            curr.pop()
            idx.pop()


backtrack([], [])

for res in result:
    for r in res:
        print(r, end=" ")
    print()

File: 백준/Silver/15663. N과 M （9）/N과 M （9）.swift
import Foundation

let input = readLine()!.split(separator: " ").map { Int(String($0))! }
let n = input[0]
let m = input[1]
let nums = readLine()!.split(separator: " ").map { Int(String($0))! }.sorted()
var visited: [Bool] = .init(repeating: false, count: n)
var result: [Int] = .init(repeating: 0, count: m)
var answer = ""


func recursive(k: Int) {
  if k == m {
    answer += result.map { String($0) }.joined(separator: " ") + "\n"
    return
  }

  var used = 0

  for i in 0..<n {
    if !visited[i] {
      if used == nums[i] { continue }
      visited[i] = true
      result[k] = nums[i]
      used = nums[i]
      recursive(k: k + 1)
      visited[i] = false
    }
  }
}

recursive(k: 0)
print(answer)